datafusion/optimizer/src/push_down_projection.rs

```diff
@@ -523,13 +523,24 @@ fn push_down_scan(
         }
     }
 
+    // Building new projection from BTreeSet
+    // preserving source projection order if it exists
+    let projection = if let Some(original_projection) = &scan.projection {
+        original_projection
```

Comment:
What about columns that are in `projection` but not in `original_projection` (as in the original projection did not contain sufficient columns for the query)?

Reply:
After an attempt of writing test for this case I've realized that maybe we don't need additional plan validity checks in optimizer rules -- normally plan builder won't allow to project/filter/etc on missing column 🤔